float32 cup_distance
float32 cup_angle